OrbitLinuxTracing/GpuTracepointEventProcessorTest.cpp

```diff
@@ -0,0 +1,589 @@
+  event.ring_buffer_record.sample_id.time = timestamp_ns;
+  reinterpret_cast<amdgpu_sched_run_job_tracepoint*>(event.tracepoint_data.get())->context =
+      context;
+  EXPECT_EQ(event.GetContext(), context);
+  reinterpret_cast<amdgpu_sched_run_job_tracepoint*>(event.tracepoint_data.get())->seqno = seqno;
+  EXPECT_EQ(event.GetSeqno(), seqno);
+  reinterpret_cast<amdgpu_sched_run_job_tracepoint*>(event.tracepoint_data.get())->timeline =
+      ((timeline.length() + 1) << 16) | sizeof(amdgpu_sched_run_job_tracepoint);
+  memcpy(event.tracepoint_data.get() + sizeof(amdgpu_sched_run_job_tracepoint), timeline.c_str(),
+         timeline.length() + 1);
+  EXPECT_EQ(event.ExtractTimelineString(), timeline);
+  return event;
+}
+
+DmaFenceSignaledPerfEvent MakeFakeDmaFenceSignaledPerfEvent(uint64_t timestamp_ns, uint32_t context,
+                                                            uint32_t seqno,
+                                                            const std::string& timeline) {
+  DmaFenceSignaledPerfEvent event{
+      static_cast<uint32_t>(sizeof(dma_fence_signaled_tracepoint) + timeline.length() + 1)};
+  event.ring_buffer_record.sample_id.time = timestamp_ns;
+  reinterpret_cast<dma_fence_signaled_tracepoint*>(event.tracepoint_data.get())->context = context;
+  EXPECT_EQ(event.GetContext(), context);
+  reinterpret_cast<dma_fence_signaled_tracepoint*>(event.tracepoint_data.get())->seqno = seqno;
+  EXPECT_EQ(event.GetSeqno(), seqno);
+  reinterpret_cast<dma_fence_signaled_tracepoint*>(event.tracepoint_data.get())->timeline =
+      ((timeline.length() + 1) << 16) | sizeof(dma_fence_signaled_tracepoint);
+  memcpy(event.tracepoint_data.get() + sizeof(dma_fence_signaled_tracepoint), timeline.c_str(),
+         timeline.length() + 1);
+  EXPECT_EQ(event.ExtractTimelineString(), timeline);
+  return event;
+}
+
+orbit_grpc_protos::GpuJob MakeGpuJob(int32_t tid, uint32_t context, uint32_t seqno,
+                                     std::string timeline, int32_t depth,
+                                     uint64_t amdgpu_cs_ioctl_time_ns,
+                                     uint64_t amdgpu_sched_run_job_time_ns,
+                                     uint64_t gpu_hardware_start_time_ns,
+                                     uint64_t dma_fence_signaled_time_ns) {
+  orbit_grpc_protos::GpuJob expected_gpu_job;
+  expected_gpu_job.set_tid(tid);
+  expected_gpu_job.set_context(context);
+  expected_gpu_job.set_seqno(seqno);
+  expected_gpu_job.set_timeline(std::move(timeline));
+  expected_gpu_job.set_depth(depth);
+  expected_gpu_job.set_amdgpu_cs_ioctl_time_ns(amdgpu_cs_ioctl_time_ns);
+  expected_gpu_job.set_amdgpu_sched_run_job_time_ns(amdgpu_sched_run_job_time_ns);
+  expected_gpu_job.set_gpu_hardware_start_time_ns(gpu_hardware_start_time_ns);
+  expected_gpu_job.set_dma_fence_signaled_time_ns(dma_fence_signaled_time_ns);
+  return expected_gpu_job;
+}
+
+void ExpectGpuJobEq(const orbit_grpc_protos::GpuJob& actual,
```

Comment:
I am fine with this. But have you considered making this method a matcher? Might make it even clearer on the caller.

Reply:
Yes, I had originally written it as a matcher. Since you don't have the individual `EXPECT_EQ`s the output in case of failure is completely useless, because you don't know where the difference is.

Reply:
You could use the `AllOf` and `Fields`/`Property` Matcher. That should preserve the output.
See [here](https://chromium.googlesource.com/external/github.com/google/googletest/+/HEAD/googlemock/docs/cook_book.md#validating-a-member-of-an-object).

```
Matcher<GpuJob> GpuJobEq(
    const GpuJob& expected) {
  return AllOf(
      Property(&GpuJob:: tid, expected. tid()),
      Field(&GpuJob:: context, expected. context()));
}
```

Reply:
Interesting, I'll have a look. Then I use the matcher in the `EXPECT_CALL` directly?

Reply:
Either in the `EXPECT_CALL` directly, or via `EXPECT_THAT`.

Reply:
So I'm trying with the matcher and I kind of have the same problem, that is the output is horrible and it doesn't tell me what field is wrong, just that no call was registered as the parameter didn't match.

For example, by making a test fail on purpose:
```
unknown file: Failure

Unexpected mock function call - returning directly.
    Function call: OnGpuJob(88-byte object <A8-75 22-7D 1D-56 00-00 00-00 00-00 00-00 00-00 00-00 00-00 2A-00 00-00 01-00 00-00 0A-00 00-00 64-00 00-00 00-00 00-00 C8-00 00-00 00-00 00-00 C8-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 2C-01 00-00 00-00 00-00 B0-54 D1-7D 1D-56 00-00 00-00 00-00 05-00 00-00>)
Google Mock tried the following 1 expectation, but it didn't match:

/usr/local/google/home/dpallotti/CLionProjects/orbitprofiler/OrbitLinuxTracing/GpuTracepointEventProcessorTest.cpp:166: EXPECT_CALL(mock_listener_, OnGpuJob(GpuJobEq(expected_gpu_job)))...
  Expected arg #0: (is an object whose given property is equal to 42) and (is an object whose given property is equal to 1) and (is an object whose given property is equal to 10) and (is an object whose given property is equal to "timeline") and (is an object whose given property is equal to 1) and (is an object whose given property is equal to 100) and (is an object whose given property is equal to 200) and (is an object whose given property is equal to 200) and (is an object whose given property is equal to 300)
           Actual: 88-byte object <A8-75 22-7D 1D-56 00-00 00-00 00-00 00-00 00-00 00-00 00-00 2A-00 00-00 01-00 00-00 0A-00 00-00 64-00 00-00 00-00 00-00 C8-00 00-00 00-00 00-00 C8-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 2C-01 00-00 00-00 00-00 B0-54 D1-7D 1D-56 00-00 00-00 00-00 05-00 00-00>, whose given property is 0 (of type int)
         Expected: to be called once
           Actual: never called - unsatisfied and active
/usr/local/google/home/dpallotti/CLionProjects/orbitprofiler/OrbitLinuxTracing/GpuTracepointEventProcessorTest.cpp:166: Failure
Actual function call count doesn't match EXPECT_CALL(mock_listener_, OnGpuJob(GpuJobEq(expected_gpu_job)))...
         Expected: to be called once
           Actual: never called - unsatisfied and active
```

vs a much clearer

```
/usr/local/google/home/dpallotti/CLionProjects/orbitprofiler/OrbitLinuxTracing/GpuTracepointEventProcessorTest.cpp:123: Failure
Expected equality of these values:
  actual.depth()
    Which is: 0
  expected.depth()
    Which is: 1
```

Reply:
Ok, I guess I need to use `EXPECT_THAT` then. I feel like that doesn't really add anything to what was there before, but as I'm already here...

Reply:
Example output:
```
/usr/local/google/home/dpallotti/CLionProjects/orbitprofiler/OrbitLinuxTracing/GpuTracepointEventProcessorTest.cpp:173: Failure
Value of: actual_gpu_job
Expected: (is an object whose given property is equal to 42) and (is an object whose given property is equal to 1) and (is an object whose given property is equal to 10) and (is an object whose given property is equal to "timeline") and (is an object whose given property is equal to 1) and (is an object whose given property is equal to 100) and (is an object whose given property is equal to 200) and (is an object whose given property is equal to 200) and (is an object whose given property is equal to 300)
  Actual: 88-byte object <A8-55 09-54 6A-55 00-00 00-00 00-00 00-00 00-00 00-00 00-00 2A-00 00-00 01-00 00-00 0A-00 00-00 64-00 00-00 00-00 00-00 C8-00 00-00 00-00 00-00 C8-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 2C-01 00-00 00-00 00-00 A0-39 4D-55 6A-55 00-00 00-00 00-00 05-00 00-00> (of type orbit_grpc_protos::GpuJob), whose given property is 0 (of type int)
```

So again, it doesn't tell me what field is different. The advantage though is that it tells me the correct line in the source code. So I don't know, pros and cons for both solutions.

Reply:
Yeah, probably the EXPECT_CALL will not tell you this, right. You can specify Strings for the `Field` and `Property` matchers. When I wrote the suggestion, I was mostly thinking of not testing the order of events, using `UnorderedElementsAre`. But as that won't be used anyways, I agree that there is not much benefit here. 

I leave you with the decision.

Reply:
I'm keeping the `EXPECT_THAT` with the matcher, it might be useful to have the matcher around. I also added the strings for the names of the properties, they add a bit of value instead of getting "given property". It would be better to be told what property causes the fail but oh well, at least the line number is correct with this.

Reply:
Sounds good.